module/validation/receipt_validator.go

```diff
@@ -147,77 +148,209 @@ func (v *receiptValidator) resultChainCheck(result *flow.ExecutionResult, prevRe
+//    block and only results from this fork are included
+//  * no duplicates in fork
+// Expected errors during normal operations:
+// * engine.InvalidInputError
+// * engine.UnverifiableInputError
+func (v *receiptValidator) ValidatePayload(candidate *flow.Block) error {
+	header := candidate.Header
+	payload := candidate.Payload
+
+	// Get the latest sealed result on this fork and the corresponding block,
+	// whose result is sealed. This block is not necessarily finalized.
+	lastSeal, err := v.seals.ByBlockID(header.ParentID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve latest seal for fork with head %x: %w", header.ParentID, err)
+	}
+	latestSealedResult, err := v.results.ByID(lastSeal.ResultID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve latest sealed result %x: %w", lastSeal.ResultID, err)
+	}
+	sealedBlock, err := v.headers.ByBlockID(lastSeal.BlockID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve sealed block (%x): %w", lastSeal.BlockID, err)
+	}
+	sealedHeight := sealedBlock.Height
+
+	// forkBlocks is the set of all _unsealed_ blocks on the fork. We
+	// use it to identify receipts that are for blocks not in the fork.
+	forkBlocks := make(map[flow.Identifier]struct{})
+
+	// Set of the execution tree with root latestSealedResult.
+	// Used for detecting duplicates and results with invalid parent results.
+	executionTree := make(map[flow.Identifier]*flow.ExecutionResult)
+	executionTree[lastSeal.ResultID] = latestSealedResult
+
+	// Set of previously included results. Used for detecting duplicates.
+	forkReceipts := make(map[flow.Identifier]struct{})
+
+	// Start from the lowest unfinalized block and walk the chain upwards until we
+	// hit the candidate's parent. For each visited block track:
+	bookKeeper := func(blockID flow.Identifier, payloadIndex *flow.Index) error {
+		// track encountered blocks
+		forkBlocks[blockID] = struct{}{}
+
+		// track encountered receipts
+		for _, recID := range payloadIndex.ReceiptIDs {
+			forkReceipts[recID] = struct{}{}
 		}
 
-		return v.previousResult(executionResult)
+		// extend execution tree
+		for _, resultID := range payloadIndex.ResultIDs {
+			result, err := v.results.ByID(resultID)
+			if err != nil {
+				return fmt.Errorf("could not retrieve result %v: %w", resultID, err)
+			}
+			if _, ok := executionTree[result.PreviousResultID]; !ok {
+				// We only collect results that directly descend from the last sealed result.
+				// Because Results are listed in an order that satisfies the parent-first
+				// relationship, we can skip all results whose parent are unknown.
+				continue
+			}
+			executionTree[resultID] = result
+		}
+		return nil
 	}
 
-	for i, r := range receipts {
-		err := v.validate(r, previousResult)
+	// Demonstration that order of block traversal is critical for correctness of algorithm:
+	err = v.forkCrawler(header.ParentID, sealedHeight+1, bookKeeper) // crawling bottom up works
```

Comment:
I think we might be able to optimize further by traversing less blocks (optional for now)

The idea is based on the two facts:
1) the parent block must have been validated already
2) the result or receipts will only be included at later blocks than the executed block. 

Let's see an example: it's a happy path where each block includes two receipts for it's parent block and a seal for it's grand parent block
```
A -> B (ER_A_1, ER_A_2) -> C (ER_B_1, ER_B_2, Seal_A) -> D (ER_C_1, ER_C_2, Seal_B) (Finalized) -> E (ER_D_1, ER_D_2, Seal_C) -> F (ER_E_1, ER_E_2, Seal_D) -> G (ER_F_1, ER_F_2, Seal_E)
```

We are validating block G. The current logic will traverse from `D` (because G's parent includes a Seal_D) to `F`.

But since block G only includes results for block F, we could start from `F` and skip D and E. 

If a receipt is referring a result which can be found in the same payload, then we are able to validate it (including duplication check) with backwards traversing and can stop at the block of that result.

If a receipt is referring a result which can't be found in the same payload (regardless it's on a different fork, or it's own fork), we are still able to validate with backwards traversing. 

Here is the pseudo code:
```
newExecutedBlockIDs, err := allBlockIDsFromResults(payload.Results) // error if there is duplicated result
// previous receipts are receipts that referring a result are not included in the same block payload
// invalid if there is duplicated receipt
previousReceipts := findPreviousReceipts(payload.Results, payload.Receipts)
resultIDs := getResultIDs(previousReceipts)
// invalid if result is not found in storage
results := findResultsFromStorageByID(resultIDs)
// must be found, because if we have stored the result, then we must stored the executed block
previousExecutedBlockIDs := findBlockForResults(results)

lowest := findLowestBlock(append(newExecutedBlockIDs, previousExecutedBlockIDs...) // we will only traverse up to this block

traverseForwardFromTo(lowest, candidate.Header.ParentID, func(block) {
  // the rest are the same
 // ensure the results are forms a tree
 // ensure the receipts are unique
})

```

Reply:
Extracted into separate issue. https://github.com/dapperlabs/flow-go/issues/5402